Make graphics render check more flexible (bugfix)

rickwu666666 opened the pull request:
<!--
Example Title: Fixed bugged behaviour of checkbox load config (Bugfix)

A Traceability Marker is required as a suffix in the PR title to help understand the impact of your change at a glance.

Pick one of the following:
- Infra: Your change only includes documentation, comments, github actions or metabox
- BugFix: Your change fixes a bug
- New: Your change is a new backward compatible feature, a new test/test plan/test inclusion
- Breaking: Your change breaks backward compatibility.
    - This includes any API change to checkbox-ng/checkbox-support
    - Changes to PXU grammar/field requirements
    - Breaking changes to dependencies in snaps (fwts upgrade for example)

If your change is to providers it can only be (Infra, BugFix or New).

If your change impacts the submission format in Checkbox test reports, ensure that `submission-schema/schema.json` is updated and relevant fields are documented.

Signed commits are required.
  - See CONTRIBUTING.md (https://github.com/canonical/checkbox/blob/main/CONTRIBUTING.md#signed-commits-required) for further instructions.
  - If you are posting your first pull request from a fork of the repository, a Checkbox maintainer (someone with contributor / maintainer / admin rights) will be required to enable CI checks in the repo to be executed.
    - This will be communicated with a comment to the PR of the form `/canonical/self-hosted-runners/run-workflows <SHA-for-HEAD-commit>`
-->

## Description
Rewrite the script in python and make the matching logic more flexible for GL_VENDOR and GL_RENDERER, since in some case it will report the GL_RENDERER with the version behind but we care about only the renderer it self.
<!--
Describe your changes here:

- What's the problem solved (briefly, since the issue is where this is elaborated in more detail).
- Introduce your implementation approach in a way that helps reviewing it well.
- Alert the reviewer of any changes that involve data persistence: present examples of file format changes as part of the PR description (e.g. new fields of data stored in unit or submission output).
-->

## Resolved issues

<!--
Note the Jira and GitHub issue(s) resolved by this PR (`Fixes|Resolves ...`).
Make sure that the linked issue titles & descriptions are also up to date.
-->

## Documentation

<!--
Please make sure that...
- Documentation impacted by the changes is up to date (becomes so, remains so).
  - Documentation in the repository, including contribution guidelines.
  - Process documentation outside the repository.
- Tests are included for the changed functionality in this PR. If to be merged without tests, please elaborate why.
- When breaking changes and other key changes are introduced, the PR having been merged should be broadcast (in demo sessions, IM, Discourse) with relevant references to documentation. This is an opportunity to gather feedback and confirm that the changes and how they are documented are understood.
-->

## Tests
Sideload result
```
=========[ Running job 1 / 1. Estimated time left (at least): 0:00:00 ]=========                                                                                                        
---------------[ Run OpenGL ES 2.0 Wayland benchmark on the GPU ]---------------                                                                                                        
ID: com.canonical.contrib::mir/glmark2-es2-wayland-auto                                                                                                                                 
Category: com.canonical.contrib::mir_test                                                                                                                                                                                             
... 8< -------------------------------------------------------------------------                                                                                                                                                      
2025-12-09 05:41:56,575 - INFO - The ubuntu-frame is active                                                                                                                                                                                                           
2025-12-09 05:41:56,577 - INFO - Running glmark2-es2-wayland benchmark...                                                                                                                                                                                             
2025-12-09 05:41:57,821 - INFO - WARNING: x11 interface not connected! Please run: /snap/graphics-test-tools/current/bin/setup.sh                                                                                                                                     
2025-12-09 05:41:57,999 - INFO - Could not open module param file '/sys/module/mali_kbase/parameters/large_page_conf'                                                                                                                                                 
2025-12-09 05:41:58,100 - INFO - =======================================================                                                                                                                                                                              
2025-12-09 05:41:58,104 - INFO - glmark2 2023.01                                                                                                                                        
2025-12-09 05:41:58,105 - INFO - =======================================================                                                                                                                                                                                                                          
2025-12-09 05:41:58,105 - INFO - OpenGL Information                                                                                                                                                                                   
2025-12-09 05:41:58,115 - INFO - GL_VENDOR:      ARM                                                                                                                                                                                                                                                              
2025-12-09 05:41:58,119 - INFO - GL_RENDERER:    Mali-G31                                                                                                                                                                                                                                                         
2025-12-09 05:41:58,123 - INFO - GL_VERSION:     OpenGL ES 3.2 v1.r50p0-00eac0.9e625e4a80ebab2bc0b93f0394937ef0                                                                                                                                                                                                   
2025-12-09 05:41:58,125 - INFO - Surface Config: buf=32 r=8 g=8 b=8 a=8 depth=24 stencil=0 samples=0                                                                                    
2025-12-09 05:41:58,135 - INFO - Surface Size:   800x600 windowed                                                                                                                                                                                                                                                 
2025-12-09 05:41:58,140 - INFO - =======================================================                                                                                                                                              
2025-12-09 05:42:08,352 - INFO - [build] use-vbo=false: FPS: 114 FrameTime: 8.839 ms                                                                                                    
2025-12-09 05:42:18,484 - INFO - [build] use-vbo=true: FPS: 121 FrameTime: 8.299 ms                                                                                                                                                   
2025-12-09 05:42:28,631 - INFO - [texture] texture-filter=nearest: FPS: 117 FrameTime: 8.571 ms                                                                                         
2025-12-09 05:42:38,749 - INFO - [texture] texture-filter=linear: FPS: 120 FrameTime: 8.401 ms                                                                                                                                        
2025-12-09 05:42:48,849 - INFO - [texture] texture-filter=mipmap: FPS: 127 FrameTime: 7.921 ms                                                                                                                                        
2025-12-09 05:42:58,979 - INFO - [shading] shading=gouraud: FPS: 110 FrameTime: 9.097 ms                                                                                                                                                                              
2025-12-09 05:43:09,168 - INFO - [shading] shading=blinn-phong-inf: FPS: 108 FrameTime: 9.299 ms                                                                                                                                                                      
2025-12-09 05:43:19,330 - INFO - [shading] shading=phong: FPS: 99 FrameTime: 10.106 ms                                                                                                                                                                                
2025-12-09 05:43:29,530 - INFO - [shading] shading=cel: FPS: 96 FrameTime: 10.496 ms                                                                                                                                                                                  
2025-12-09 05:43:39,838 - INFO - [bump] bump-render=high-poly: FPS: 72 FrameTime: 13.961 ms                                                                                                                                           
2025-12-09 05:43:50,042 - INFO - [bump] bump-render=normals: FPS: 133 FrameTime: 7.546 ms                                                                                                                                             
2025-12-09 05:44:00,280 - INFO - [bump] bump-render=height: FPS: 121 FrameTime: 8.299 ms                                                                                                                                              
2025-12-09 05:44:10,388 - INFO - [effect2d] kernel=0,1,0;1,-4,1;0,1,0;: FPS: 104 FrameTime: 9.623 ms                                                                                                                                                                                                              
2025-12-09 05:44:20,513 - INFO - [effect2d] kernel=1,1,1,1,1;1,1,1,1,1;1,1,1,1,1;: FPS: 56 FrameTime: 17.938 ms                                                                                                                                                                                                   
2025-12-09 05:44:30,604 - INFO - [pulsar] light=false:quads=5:texture=false: FPS: 123 FrameTime: 8.157 ms                                                                                                                                                                                                         
2025-12-09 05:44:41,400 - INFO - [desktop] blur-radius=5:effect=blur:passes=1:separable=true:windows=4: FPS: 42 FrameTime: 24.089 ms                                                                                                                                                                              
2025-12-09 05:44:51,542 - INFO - [desktop] effect=shadow:windows=4: FPS: 88 FrameTime: 11.404 ms                                                                                                                                      
2025-12-09 05:45:01,802 - INFO - [buffer] columns=200:interleave=false:update-dispersion=0.9:update-fraction=0.5:update-method=map: FPS: 38 FrameTime: 26.858 ms                                                                                                                                                  
2025-12-09 05:45:12,087 - INFO - [buffer] columns=200:interleave=false:update-dispersion=0.9:update-fraction=0.5:update-method=subdata: FPS: 38 FrameTime: 26.591 ms                                                                                                                                              
2025-12-09 05:45:22,330 - INFO - [buffer] columns=200:interleave=true:update-dispersion=0.9:update-fraction=0.5:update-method=map: FPS: 54 FrameTime: 18.652 ms                                                                                                                                                                                                                 
2025-12-09 05:45:33,546 - INFO - [ideas] speed=duration: FPS: 62 FrameTime: 16.182 ms                                                                                                                                                                                                                                                                                           
2025-12-09 05:45:44,333 - INFO - [jellyfish] <default>: FPS: 74 FrameTime: 13.538 ms                                                                                                    
2025-12-09 05:45:56,546 - INFO - [terrain] <default>: FPS: 8 FrameTime: 134.304 ms                                                                                                      
2025-12-09 05:46:07,095 - INFO - [shadow] <default>: FPS: 69 FrameTime: 14.606 ms                                                                                                                                                                                                                                 
2025-12-09 05:46:19,897 - INFO - [refract] <default>: FPS: 19 FrameTime: 53.504 ms                                                                                                                                                    
2025-12-09 05:46:30,032 - INFO - [conditionals] fragment-steps=0:vertex-steps=0: FPS: 125 FrameTime: 8.046 ms                                                                                                                                                                                                     
2025-12-09 05:46:40,147 - INFO - [conditionals] fragment-steps=5:vertex-steps=0: FPS: 97 FrameTime: 10.340 ms                                                                                                                                                                                                     
2025-12-09 05:46:50,276 - INFO - [conditionals] fragment-steps=0:vertex-steps=5: FPS: 119 FrameTime: 8.432 ms                                                                                                                                                                                                     
2025-12-09 05:47:00,409 - INFO - [function] fragment-complexity=low:fragment-steps=5: FPS: 115 FrameTime: 8.757 ms                                                                                                                                                                                                
2025-12-09 05:47:10,540 - INFO - [function] fragment-complexity=medium:fragment-steps=5: FPS: 90 FrameTime: 11.197 ms                                                                                                                                                                                             
2025-12-09 05:47:20,681 - INFO - [loop] fragment-loop=false:fragment-steps=5:vertex-steps=5: FPS: 113 FrameTime: 8.892 ms                                                                                                                                                                                                                                                                                                                                                   
2025-12-09 05:47:30,801 - INFO - [loop] fragment-steps=5:fragment-uniform=false:vertex-steps=5: FPS: 116 FrameTime: 8.688 ms                                                                                                                                                                                                                                                                                                                                                
2025-12-09 05:47:40,960 - INFO - [loop] fragment-steps=5:fragment-uniform=true:vertex-steps=5: FPS: 98 FrameTime: 10.282 ms                                                                                                           
2025-12-09 05:47:40,967 - INFO - =======================================================                                                                                                                                              
2025-12-09 05:47:40,978 - INFO - glmark2 Score: 89                                                                                                       
2025-12-09 05:47:40,981 - INFO - =======================================================                                                                 
2025-12-09 05:47:42,132 - INFO - PASS: GL_VENDOR is 'AR'                                                                                                 
2025-12-09 05:47:42,136 - INFO - PASS: GL_RENDERER is 'Mali'                                                                                                                                                                                                                                                      
------------------------------------------------------------------------- >8 ---                                                                                                                                                                                                                                  
Outcome: job passed    
```
<!--
- How was this PR tested? Please provide steps to follow so that the reviewer(s) can test on their end.
- Please provide a list of what tests were run and on what platform/configuration.
- Remember to check the test coverage of your PR as described in CONTRIBUTING.md
-->

rickwu666666 committed:
Rewrite shell with python

rickwu666666 committed:
use format instead of f string

rickwu666666 committed:
Add unittests

rickwu666666 committed:
Modify execute from shell to python in test jobs

rickwu666666 committed:
Fix black format for test_graphics_test.py

rickwu666666 committed:
Fix for py3.5

rickwu666666 committed:
replace assert_called_once with call_count for py3.5 check

stanley31huang commented on contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py:
```diff
@@ -0,0 +1,158 @@
+
+def main():
+    """Main function."""
+    setup_logging()
```
there's a verbose variable in setup_logging function to adjust the log level, thus I would suggest we add a argument likes `--debug`.

rickwu666666 committed:
Use argparse to handle the input args

rickwu666666 committed:
Modify unittest for the code change

tomli380576 commented on contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py:
```diff
@@ -0,0 +1,157 @@
+            "FAIL: 'GL_VENDOR' or 'GL_RENDERER' is empty. "
+            "Please set them in config file!"
+        )
+        return 1
```
Optional: I think these can be moved to the beginning of this test so you don't have to wait for the benchmark to finish to check their existence

tomli380576 commented on contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py:
```diff
@@ -0,0 +1,157 @@
+        logger.info(line.strip())
+        output += line
+
+    process.wait()
```
I think you can directly call `subprocess.check_output` here since we block the main thread until glmark2 is done

rickwu666666 committed:
Move config variables check at the first

Use subprocess.check_output instead of Popen. This change will cause the test log only present after glmark2 test finished. However, it should be fine since this is a automation job

rickwu666666 committed:
Modify unittest for the code change

rickwu666666 committed:
Fix black format

stanley31huang commented on contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py:
```diff
@@ -0,0 +1,152 @@
+
+
+def test_ubuntu_frame_launching():
+    """Test ubuntu-frame launching."""
```
I would suggest we change the implementation with python subprocess.popen, please refer the details as following
```
import subprocess
import time


def check_process_is_alive(proc):
    if proc.poll() is None:
        print(f"Process is alive... pid: {proc.pid}")
    else:
        print(f"Process ended unexpectedly with return code {proc.returncode}")
        return False
    return True

cmd = ['top', '-d', '1']

print(f"Launching command: {' '.join(cmd)}")

    # stdout=subprocess.DEVNULL suppresses the massive output top generates
    # Remove stdout argument if you want to see the output in your console
process = subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.PIPE)

print(f"Process started with PID: {process.pid}")
print("-" * 30)

try:
    # 2. Loop for 30 seconds
    duration = 30
    for i in range(duration):
            # 3. Check if process is alive
            # process.poll() returns None if the process is still running

        if not check_process_is_alive(process):
            break
        time.sleep(1)

    if check_process_is_alive(process):
        print("kill process by scripts...")
        process.terminate()
    else:
        raise SystemExit("1")

except Exception as e:
    print(e)
    ```

rickwu666666 committed:
Update test_ubuntu_frame_launching to not use timeout instead of subprocess.popen

rickwu666666 committed:
Modify unittest for code change

rickwu666666 committed:
Fix black format

rickwu666666 committed:
Fix unittest for pass py3.5

stanley31huang commented on contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py:
```diff
@@ -0,0 +1,156 @@
+                    proc.returncode,
+                )
+                return 1
+    return 0
```
```suggestion
    if is_ubuntu_frame_active():
        logger.info("No need to bring it up again")
        logger.info("journal log of ubuntu frame:")
        subprocess.run(["journalctl", "-b", "0", "-g", "ubuntu-frame"])
    else:        
        try:
            proc = launch_ubuntu_frame()
            proc.wait(timeout=20)
        except subprocess.TimeoutExpired:
            proc.terminate()
            proc.wait()
            logger.info("PASS: Timeout reached without any failures detected.")
        else:
            if proc.returncode != 0:
                logger.error(
                    "FAIL: ubuntu-frame exited earlier with code %s",
                    proc.returncode,
                )
                return 1
    return 0


def launch_ubuntu_frame():
    logger.info("Activating ubuntu-frame now...")
    proc = subprocess.Popen(
        ["ubuntu-frame"],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    return proc
```

stanley31huang commented on contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py:
```diff
@@ -0,0 +1,156 @@
+    logger.info(output)
+
+    if frame_pid:
+        subprocess.run(["kill", str(frame_pid)])
```
```suggestion
    if proc and proc.pid:
        proc.terminate()
```

rickwu666666 committed:
Update contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py

Co-authored-by: stanley31huang <stanley.huang@canonical.com>

rickwu666666 committed:
Update contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py

Co-authored-by: stanley31huang <stanley.huang@canonical.com>

rickwu666666 committed:
Update contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py

Co-authored-by: stanley31huang <stanley.huang@canonical.com>

rickwu666666 committed:
Fix black format

rickwu666666 committed:
Remove unuse variable

rickwu666666 committed:
Add initial value for proc in case script crash if ubuntu frame running by default

rickwu666666 committed:
Modify unittests

rickwu666666 committed:
Fix unittest for py3.5

stanley31huang commented on contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py:
```diff
@@ -0,0 +1,155 @@
+        )
+        exit_code = 1
+    else:
+        logger.info("PASS: GL_VENDOR is '{}'".format(gl_vendor))
```
I would suggest print out the actual GL_VENDOR here, not the value provided by user

stanley31huang reviewed (changes requested):
please see my inline comments

rickwu666666 committed:
Update contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py

Co-authored-by: stanley31huang <stanley.huang@canonical.com>

rickwu666666 committed:
Update contrib/checkbox-ce-oem/checkbox-provider-ce-oem/bin/graphics_test.py

Co-authored-by: stanley31huang <stanley.huang@canonical.com>

rickwu666666 committed:
print actual vendor and renderer if matched

rickwu666666 committed:
update unittest

rickwu666666 committed:
Add logic to check if GL_VENDOR and GL_RENDERER eisxt

tomli380576 reviewed (dismissed):
lgtm just a minor ordering issue

rickwu666666 committed:
Adjust the command order

tomli380576 reviewed (approved):
lgtm

stanley31huang reviewed (approved):
LGTM
